src/ringct/rctSigs.cpp

```diff
@@ -1688,51 +1691,80 @@ namespace rct {
       return true;
     }
 
-    bool batchVerifyFcmpPpProofs(std::vector<fcmp_pp::FcmpPpVerifyInput> &&fcmp_pp_verify_inputs) {
+    bool batchVerifyFcmpPpProofs(std::vector<fcmp_pp::FcmpPpVerifyInput> &&fcmp_pp_verify_inputs,
+        const std::vector<std::size_t> & n_inputs_per_proof) {
       const std::size_t n_proofs = fcmp_pp_verify_inputs.size();
+      CHECK_AND_ASSERT_MES(n_proofs == n_inputs_per_proof.size(), false, "Did not have matching n inputs per proof");
+      if (n_proofs == 0)
+          return true;
+
+      // Sort the inputs in ascending order based on input count. Smaller input counts will be batched together.
+      std::size_t total_n_inputs = 0;
+      std::vector<std::pair<fcmp_pp::FcmpPpVerifyInput, std::size_t>> verify_inputs;
+      verify_inputs.reserve(n_proofs);
+      for (std::size_t i = 0; i < n_proofs; ++i)
+      {
+        total_n_inputs += n_inputs_per_proof.at(i);
+        verify_inputs.emplace_back(std::make_pair(std::move(fcmp_pp_verify_inputs.at(i)), n_inputs_per_proof.at(i)));
+      }
+      std::sort(verify_inputs.begin(), verify_inputs.end(), [](auto &a, auto &b) { return a.second < b.second; });
 
       tools::threadpool &tpool = tools::threadpool::getInstanceForCompute();
       tools::threadpool::waiter waiter(tpool);
       const std::size_t n_threads = std::max<std::size_t>(1, tpool.get_max_concurrency());
       const bool multithreaded = n_threads > 1;
 
-      std::vector<std::vector<fcmp_pp::FcmpPpVerifyInput>> batches;
-      batches.reserve(n_proofs); // over reserve
+      // 1 batch per thread
+      const std::size_t n_batches = n_threads;
 
-      // Split batches across all available threads
-      std::size_t sanity_counter = 0;
-      const std::size_t fcmp_pp_verify_batch_size = std::max<std::size_t>(1, (n_proofs / n_threads));
-      for (std::size_t i = 0; i < n_proofs; i += fcmp_pp_verify_batch_size)
+      std::vector<std::vector<fcmp_pp::FcmpPpVerifyInput>> batches;
```

Comment:
You can remove `1 + n_batches` allocations and the overhead of the `std::unique_ptr` deallocs by passing indices of `verify_inputs` to the jobs instead.